rust/src/history.rs

```diff
@@ -0,0 +1,783 @@
+    first_boot_timestamp: u64,
+    /// The last time the deployment was booted if multiple consecutive times.
+    last_boot_timestamp: u64,
+    /// `true` if there are no more entries.
+    eof: bool,
+}
+
+impl HistoryEntry {
+    /// Create a new `HistoryEntry` from a boot marker and a deployment marker.
+    fn new_from_markers(boot: BootMarker, deploy: DeploymentMarker) -> HistoryEntry {
+        HistoryEntry {
+            first_boot_timestamp: boot.timestamp,
+            last_boot_timestamp: boot.timestamp,
+            deploy_timestamp: deploy.timestamp,
+            deploy_cmdline: deploy
+                .cmdline
+                .map(|s| s.into_raw())
+                .unwrap_or(ptr::null_mut()),
+            boot_count: 1,
+            eof: false,
+        }
+    }
+
+    fn eof() -> HistoryEntry {
+        HistoryEntry {
+            eof: true,
+            first_boot_timestamp: 0,
+            last_boot_timestamp: 0,
+            deploy_timestamp: 0,
+            deploy_cmdline: ptr::null_mut(),
+            boot_count: 0,
+        }
+    }
+}
+
+#[derive(PartialEq)]
+enum JournalSearchMode {
+    BootMsgs,
+    BootAndDeploymentMsgs,
+}
+
+#[cfg(not(test))]
+fn journal_record_timestamp(journal: &journal::Journal) -> Fallible<u64> {
+    Ok(journal
+        .timestamp()?
+        .duration_since(std::time::UNIX_EPOCH)?
+        .as_secs())
+}
+
+#[cfg(test)]
+fn journal_record_timestamp(journal: &journal::Journal) -> Fallible<u64> {
+    Ok(journal.current_timestamp.unwrap())
+}
+
+fn map_to_u64<T>(s: Option<&T>) -> Option<u64>
+where
+    T: Deref<Target = str>,
+{
+    s.and_then(|s| s.parse::<u64>().ok())
+}
+
+fn history_get_oldest_deployment_msg_timestamp() -> Fallible<Option<u64>> {
+    let mut journal = journal::Journal::open(journal::JournalFiles::System, false, true)?;
+    journal.seek(journal::JournalSeek::Head)?;
+    journal.match_add("MESSAGE_ID", RPMOSTREE_DEPLOY_MSG)?;
+    while let Some(rec) = journal.next_record()? {
+        if let Some(ts) = map_to_u64(rec.get("DEPLOYMENT_TIMESTAMP")) {
+            return Ok(Some(ts));
+        }
+    }
+    Ok(None)
+}
+
+/// Gets the oldest deployment message in the journal, and nuke all the GVariant data files
+/// that correspond to deployments older than that one. Essentially, this binds pruning to
+/// journal pruning. Called from C through `ror_history_prune()`.
+fn history_prune() -> Fallible<()> {
+    let oldest_timestamp = history_get_oldest_deployment_msg_timestamp()?;
+
+    // Cleanup any entry older than the oldest entry in the journal. Also nuke anything else that
+    // doesn't belong here; we own this dir.
+    let dir = Dir::open(RPMOSTREE_HISTORY_DIR)?;
+    for entry in dir.list_dir(".")? {
+        let entry = entry?;
+        let ftype = match entry.simple_type() {
+            Some(ftype) => ftype,
+            /* stat() fallback */
+            None => dir.metadata(entry.file_name())?.simple_type(),
+        };
```

Comment:
This is also `get_file_type()` in `OpenatDirExt`.  (But eh, don't need to port to openat if you don't want to)